lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,783 @@
+#include "seahorn/HornClauseDB.hh"
+
+#include "llvm/IR/Function.h"
+#include "llvm/Support/CommandLine.h"
+
+#include "ufo/Expr.hpp"
+#include "ufo/Smt/Z3n.hpp"
+#include "ufo/Smt/EZ3.hh"
+#include <vector>
+#include <boost/logic/tribool.hpp>
+#include "seahorn/HornClauseDBWto.hh"
+#include <algorithm>
+
+#include "ufo/Stats.hh"
+
+using namespace llvm;
+
+namespace seahorn
+{
+	char PredicateAbstraction::ID = 0;
+
+	bool PredicateAbstraction::runOnModule (Module &M)
+	{
+		HornifyModule &hm = getAnalysis<HornifyModule> ();
+
+		PredicateAbstractionAnalysis pabs(hm);
+		pabs.runAnalysis();
+
+		return false;
+	}
+
+	void PredicateAbstraction::getAnalysisUsage (AnalysisUsage &AU) const
+	{
+		AU.addRequired<HornifyModule> ();
+		AU.setPreservesAll();
+	}
+
+	void PredicateAbstractionAnalysis::runAnalysis()
+	{
+		Stats::resume ("Pabs solve");
+		//load the Horn clause database
+		auto &db = m_hm.getHornClauseDB ();
+		db.buildIndexes ();
+
+		//guess candidates
+		guessCandidate(db);
+
+		HornDbModel oldModel;
+		oldModel.setCurrentCandidates(currentCandidates);
+
+		PredAbsHornModelConverter converter;
+
+		//run main algorithm
+		HornClauseDB new_db = generateAbstractDB(db, converter);
+
+		//initialize spacer based on new DB
+		m_fp.reset (new ZFixedPoint<EZ3> (m_hm.getZContext ()));
+		ZFixedPoint<EZ3> &fp = *m_fp;
+		ZParams<EZ3> params (m_hm.getZContext ());
+		params.set (":engine", "spacer");
+		// -- disable slicing so that we can use cover
+		params.set (":xform.slice", false);
+		params.set (":use_heavy_mev", true);
+		params.set (":reset_obligation_queue", true);
+		params.set (":pdr.flexible_trace", false);
+		params.set (":xform.inline-linear", false);
+		params.set (":xform.inline-eager", false);
+		// -- disable utvpi. It is unstable.
+		params.set (":pdr.utvpi", false);
+		// -- disable propagate_variable_equivalences in tail_simplifier
+		params.set (":xform.tail_simplifier_pve", false);
+		params.set (":xform.subsumption_checker", true);
+//		params.set (":order_children", true);
+//		params.set (":pdr.max_num_contexts", "500");
+		fp.set (params);
+		new_db.loadZFixedPoint (fp, false);
+		boost::tribool result = fp.query ();
+
+		if (result) outs () << "sat";
+		else if (!result)
+		{
+			outs() << "unsat";
+			HornDbModel absModel;
+			absModel.initModelFromFP(new_db, fp);
+			converter.setNewToOldPredMap(newToOldPredMap);
+			converter.setAbsDB(new_db);
+			converter.convert(absModel, oldModel);
+			LOG("pabs-debug", outs() << "FINAL RESULT:\n";);
+			//Print invariants
+			printInvars(db, oldModel);
+		}
+		else outs () << "unknown";
+		outs () << "\n";
+		Stats::stop("Pabs solve");
+	}
+
+	void HornDbModel::addDef(Expr fapp, Expr def)
+	{
+		if(isOpX<TRUE>(def) || isOpX<FALSE>(def))
+		{
+			relToDefMap.insert(std::pair<Expr, Expr>(bind::fname(fapp), def));
+			return;
+		}
+
+		std::ostringstream oss;
+		oss << bind::fname(bind::fname(fapp));
+		//For AbsModel
+		if(oss.str().find("_pabs") != -1)
```

Comment:
Why is predicate abstraction mixed with `HornModel`? I don't understand this code at all :(